subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeModelTarget.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.buildtree;
+
+import org.gradle.util.Path;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Objects;
+
+/**
+ * Either a build or a project as a target for model building.
+ * <p>
+ * The target is identified by minimal information provided via Tooling API,
+ * such as the root directory of the target build.
+ */
+public class BuildTreeModelTarget {
```

Comment:
💅 I feel the design would be slightly clearer with a "sealed class hierarchy" comprised of `BuildTreeModelTarget.Build(File buildRootDir)` and `BuildTreeModelTarget.Project(File buildRootDir, String projectPath)` rather than representing one of the cases via missing information